source/common/buffer/buffer_impl.cc

```diff
@@ -266,6 +266,30 @@ void* OwnedImpl::linearize(uint32_t size) {
   return slices_.front()->data();
 }
 
+RawSlice OwnedImpl::maybeLinearize(uint32_t max_size, uint32_t desired_min_size) {
+  while (!slices_.empty() && slices_[0]->dataSize() == 0) {
+    slices_.pop_front();
+  }
+
+  if (slices_.empty()) {
+    return {nullptr, 0};
+  }
+
+  const uint64_t slice_size = std::min<uint64_t>(slices_[0]->dataSize(), max_size);
+  if (slice_size >= desired_min_size) {
+    return {slices_[0]->data(), slice_size};
+  }
+
+  // The next slice will already be of the desired size, so don't copy and
+  // return the front slice.
+  if (slices_.size() >= 2 && slices_[1]->dataSize() >= max_size) {
```

Comment:
Worth considering a generalization of this logic so we refuse to copy if we ever encounter that the next slice is larger than some copy threshold?

That way a buffer containing: {1, 1, 1, 1, 1, 16kb} would only end up copying 5 bytes when called with parameters like: linearize(16kb, 4000)